presto-main/src/main/java/com/facebook/presto/server/protocol/QueryBlockingRateLimiter.java

```diff
@@ -0,0 +1,107 @@
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.inject.Inject;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import javax.annotation.PreDestroy;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+
+/*
+ * Rate Limiting per query with token bucket
+ * Rate = rateLimitBucketMaxSize/second
+ * When having sufficient tokens, Request will be responded immediately.
+ * When not having enough tokens available, it uses the delayed processing method.
+ */
+public class QueryBlockingRateLimiter
+{
+    private final LoadingCache<QueryId, RateLimiter> rateLimiterCache;
+    private final long rateLimitBucketMaxSize;
+    private final ListeningExecutorService rateLimiterExecutorService;
+
+    private CounterStat rateLimiterTriggeredCounter = new CounterStat();
+
+    @Inject
+    public QueryBlockingRateLimiter(QueryManagerConfig queryManagerConfig)
+    {
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.rateLimitBucketMaxSize = queryManagerConfig.getRateLimitBucketMaxSize();
+        // Using a custom thread pool with size 1-10 to reduce initial thread resources
+        ExecutorService executorService = new ThreadPoolExecutor(1, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), daemonThreadsNamed("rate-limiter-listener"));
+        rateLimiterExecutorService = listeningDecorator(executorService);
+        rateLimiterCache = CacheBuilder.newBuilder().maximumSize(10000).expireAfterAccess(60, TimeUnit.MINUTES).build(CacheLoader.from(key -> RateLimiter.create(rateLimitBucketMaxSize)));
+    }
+
+    /*
+     * For accidental bug-caused DoS, we will use delayed processing method to reduce the requests, even when user do not have back-off logic implemented
+     * Optimized to avoid blocking for normal usages with TryRequire first
+     * Fall back to delayed processing method to acquire a permit, in a separate thread pool
+     * Internal guava rate limiter returns time spent sleeping to enforce rate, in seconds; 0.0 if not rate-limited, we use a future to wrap around that.
+     */
+    public ListenableFuture<Double> acquire(QueryId queryId)
+    {   // if rateLimitBucketMaxSize == 0, we disable rate limiting
+        if (rateLimitBucketMaxSize == 0) {
```

Comment:
why not use -1 instead to represent disabling?

Reply:
0 means available permit or token is 0, which is more consistent to its logical representation and easier to understand.
-1 is a little hard to make sense out of the value,

Reply:
So if it's 0, it means we don't allow any usage at all; But for disabling rate limiting, it means unlimited slots for all
Looks like they're opposite logic?

Reply:
0 ==disabled == no rate limiting == unlimited slots == old behavior
they are the same.

Reply:
> 1. Can we also add tests for `QueryBlockingRateLimiter`
> 2. Like @rschlussel suggested above, add commit message to introduce the context and functionality

I would love to, acquire() itself does not have a lot of complex logic, any suggestions on the test cases?

Reply:
> 0 ==disabled == no rate limiting == unlimited slots == old behavior they are the same. so we are returning ` immediateFuture(0.0)`, and the caller logic will immediately move on.

But rateLimitBucketMaxSize being 0 seems to indicate the bucket's max size is 0, which also means no traffic is allowed at all right? I think this is where it might be confusing

Reply:
What I have in mind is we can pass in a smaller sized `rateLimiterCache` (or if it's configurable, we can pass in the corresponding config)

And we can verify
1. The behaviors when multiple queries are calling the function multiple times at the same time, they won't affect each others 
2. After exceeding the rateLimiterCache's max size, what's the expected behavior
3. rate limit functionality itself (I understand we have added that in a more high level end point test, but here I think is a better place to test the rate limit logic actually)

Reply:
Made cache params configurable. 
To further reduce memory footprint, By Default, cache holds 1000 entries with expiration 5 minutes, 
When exceeding size limit, the oldest one will get evicted.